### PR TITLE
Trwalke/x5c claim hotfix proposal

### PR DESCRIPTION
**Pull request by trwalke**
Proposed implementation of X5C public certificate claim.
This claim is not sent by default so the user needs to opt in to send it.
Changing public API with changes to AuthenticationContext.cs Ln:763

### CODE REVIEW

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/AuthenticationContext.cs [hunk 1]
```diff
@@ -760,6 +760,23 @@ internal IWebUI CreateWebAuthenticationDialog(IPlatformParameters parameters)
                 .ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// Acquires security token from the authority and sends the x509 public certificate
+        /// </summary>
+        /// <param name="resource">Identifier of the target resource that is the recipient of the requested token.</param>
+        /// <param name="clientCertificate">The client certificate to use for token acquisition.</param>
+        /// <param name="sendX5c">Sends the x509 public certificate as a x5c claim</param>
```
Comment (jmprieur) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/AuthenticationContext.cs]: sendX5c.
I would use as a comment: "
Sends the x509 public certificate to the STS as a x5c claim. This parameter enables application developers to opt-in to easy certificates roll-up in Azure AD: in the case the certificate was renewed on the machine executing ADAL.NET, setting this parameter to true will send the certificate public key to Azure AD in addition to the certificate thumbprint, so that Azure AD rolls-over it's internal knowledge of the certificate. This avoid the azure AD application admin to have to change the application manifest in the Azure portal (or to run a powershell / CLI script)
"

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/AuthenticationContext.cs [hunk 2]
```diff
@@ -760,6 +760,23 @@ internal IWebUI CreateWebAuthenticationDialog(IPlatformParameters parameters)
                 .ConfigureAwait(false);
         }
 
+        /// <summary>
```
Comment (trwalke) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/AuthenticationContext.cs]: @vibronet @jmprieur

File: devApps/AdalCoreCLRTestApp/Program.cs [hunk 3]
```diff
@@ -55,7 +55,7 @@ static void Main(string[] args)
         {
             AuthenticationContext context = new AuthenticationContext("https://login.microsoftonline.com/common", true);
             var certificate = GetCertificateByThumbprint("<CERT_THUMBPRINT>");
-            var result = await context.AcquireTokenAsync("https://graph.windows.net", new ClientAssertionCertificate("<CLIENT_ID>", certificate));
+            var result = await context.AcquireTokenAsync("https://graph.windows.net", new ClientAssertionCertificate("<CLIENT_ID>", certificate), true);
```
Comment (henrik-me) [inline devApps/AdalCoreCLRTestApp/Program.cs]: >true [](start = 146, length = 4)

Thinking the app should include a scenario where this is false (e.g. make a request afterwards where this is false to ensure the cert is not in the request)?

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs [hunk 4]
```diff
@@ -86,11 +90,15 @@ public JsonWebToken(IClientAssertionCertificate certificate, string audience)
...
 
-        public ClientAssertion Sign(IClientAssertionCertificate credential)
+        public ClientAssertion Sign(IClientAssertionCertificate credential, bool sendX5c)
         {
+            _sendX5c = sendX5c;
+            string token;
             // Base64Url encoded header and claims
-            string token = this.Encode(credential);     
-
+            lock (LockObject)
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs]: >ock (LockObject) [](start = 13, length = 16)

why is this needed?

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs [hunk 5]
```diff
@@ -203,9 +210,36 @@ internal class JWTPayload
...
+                if (credential is ClientAssertionCertificate cert)
+                {
+                    X509CertificatePublicCertValue = Convert.ToBase64String(cert.Certificate.GetRawCertData());
+                }
+#elif NETSTANDARD1_3
+                if (credential is ClientAssertionCertificate cert)
+                {
+                    X509CertificatePublicCertValue = Convert.ToBase64String(cert.Certificate.RawData);
+                }
+#endif
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs]: can this difference be hidden in a helper/extension for each of the platforms?

**Comment by henrik-me** [comment]
pls. ensure the changes also gets to the dev branch

File: devApps/AdalCoreCLRTestApp/Program.cs (hunk 3 above)
Comment (trwalke) [reply devApps/AdalCoreCLRTestApp/Program.cs]: good point

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs (hunk 5 above)
Comment (trwalke) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs]: At the moment, I'm not sure. it may require a lot of changes.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs [hunk 6]
```diff
@@ -70,6 +71,8 @@ internal class JsonWebToken
 
         private readonly JWTPayload payload;
 
+        private bool _sendX5c = false;
```
Comment (henrik-me) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs]: not consistent with the rest of the files. pls. remove _ unless we decide that is out new norm

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs (hunk 5 above)
Comment (henrik-me) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs]: it will require a few changes.... an extension method in each of the libraries which extends the functionality of the Certificate type or you can do that as an extension to ether the type of the Certificate property  (or partial class) to the ClientAssertionCertificate. this is not a big deal, however I suggest creating a refactoring task for this one which we can include in dev, then we can work on it together

**Comment by henrik-me** [comment]
looks ok. pls. ensure this is properly tested :)

**Comment by jmprieur** [comment]
LGTM. Is it something we want to add in 3.17.4 ? in that case I guess this would be 3.18.0 as the public API changes?

**Comment by henrik-me** [comment]
We would like to get this into the next release possible. @jmprieur if you are good with the public api change then we should be good to go.

**Comment by jmprieur** [comment]
I'm good with it. @vibronet do you want to get a chance to apply a veto?

**Comment by jennyf19** [comment]
Are we adding this to 3.18 release? @jmprieur @henrik-me
